### ISSUE TITLE
Cleanup products which are not found in Merchant Center

### ISSUE THREAD

**Issue by mikkamp** [woocommerce/google-listings-and-ads#1194]
We've seen a recent Uptick in API errors  which is logged in #1181 
A lot of these errors are from the API methods productstatuses.get and products.delete and are classified as a notFound error.

This can happen when for example products have been deleted in the Merchant Center but the sites still has those products marked as synced. We store the Google ID's in the database so we need to update those ID's as soon as we know that a product no longer exists in the Merchant Center.

When we run the product sync we have a handler which handles all delete errors, and if it detects a notFound error it will call `remove_google_id` to update the ID's in the database. See: https://github.com/woocommerce/google-listings-and-ads/blob/1.10.0/src/Product/ProductSyncer.php#L299-L312

However when we call productstatuses.get we create batch requests for all ID's we have stored in the local database. If we get a notFound error we just skip to the next product, see: https://github.com/woocommerce/google-listings-and-ads/blob/1.10.0/src/MerchantCenter/MerchantStatuses.php#L318-L333
A suggestion was to call the same cleanup routine to ensure that the product is marked as unsynced since it no longer exists in the database. So when we call the product sync the next time round we won't try to delete it again.

Another suggestion:
> I think adding a cleanup or sync job that goes through all products and checks if they still exist in MC would be a good idea to decrease those `notFound` errors.

We will need to evaluate whether this will increase or decrease the current amount of requests or additional API errors.

**Comment by mikkamp** [comment]
I created a [branch](https://github.com/woocommerce/google-listings-and-ads/tree/fix/1194-cleanup-not-found-products) with a scheduled cleanup task to remove any products that are not found when we retrieve the Merchant Status. It is working to cleanup the products, but after further testing it seems we might not want to add this cleanup task.

Google products can take up to 30 minutes to become available since it takes a while for them to be processed. So if we sync a product to Google and then request a product status update (using the stored ID) we will get a notFound error because it's not available yet. If we repeat the same request a little later we do get the product status, but with the proposed change we would have already cleaned up the product (marked as unsynced).

The problem stems from the fact that we use the following logic:
1. Retrieve all Google ID's of synced products
2. Send a [ProductStatuses batch](https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/custombatch) request for a 1000 products at a time
3. Handle responses and retrieve the issues for each product and store them in our issues table
4. Repeat from step 2 for the next 1000 products

Because we are requesting statuses for specific product ID's if it's done too early it's possible that some products aren't available yet, which would result in notFound errors. The other alternative is to use the [ProductStatuses list](https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/list) request to get the statuses of all the available products. Which would avoid the notFound errors, although we did originally have that in place, but we decided to switch to batch requests.

I don't recall the full reasoning but I believe one of the reasons was making it easier to divide into batches (although pagination is available for the list endpoint). It got changed around the time that we were doing scalability testing for the statuses. However I think there were additional reasons why we switched because the statuses are also used to calculate the totals.

To summarize:
My suggestion is to switch back to using [ProductStatuses list](https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/list) to avoid notFound errors, but we would need to discuss if that has any other side effects.

**Comment by mikkamp** [comment]
As part of this issue there was also a suggestion to add a cleanup/sync process which would check the data in MC and make sure we are up to date. Based on the best practices documentation, it seems it's not intended to use the API in this way. It specifically states "[Don't use the API to regularly retrieve product information you have uploaded](https://developers.google.com/shopping-content/guides/best-practices#do-not-retrieve-product-data-regularly)".

**Comment by mikkamp** [comment]
For some history on why we decided to switch to batch requests, the PR for this is #769. There were quite a few optimizations in there but it seems the main reason to switch was to optimize for speed:

> - Retrieves MC product statuses in batches of 5000. In testing, this reduced the total retrieval time in for ~2,500 synced products from 40s to 12s
> - Also avoids paging through non-GLA products in Merchant Center.

Another comment from that PR:
> This isn't a definitive solution (moving the entire process to a scheduled action will remove any perceptible lag for users), but for now it should help with some of the super-massive stores that run into timeouts on the Product Feed page.

So maybe it's time to move to a scheduled action. When it was first implemented we didn't have the overview on the product feed page:
![image](https://user-images.githubusercontent.com/11388669/152536038-1a129ad9-8e1d-47a8-a4a2-c69d8d8acb3f.png)

Since they have to wait for the scheduled product sync to complete, it would make sense to run a scheduled status feed update after the product sync completed. Since it can take up to 30 minutes for the products to be processed we could even wait to schedule it for 30 minutes after product sync completed.
We would need to think a bit more about how often we would poll this information for a refresh. Right now it's triggered when the user refreshes/views the product feed page (with cached results ensuring it's not polled more than once an hour). With a scheduled task you wouldn't want to fetch it every hour when the user might not be viewing the page, but them having to wait till the scheduled task ran if we triggered it on page view is not ideal either.
